src/crimson/os/seastore/logstore/log_node.h

```diff
@@ -0,0 +1,1543 @@
+    SUBDEBUG(seastore_t, "index={}, key={}", _iter.index, iter->get_key());
+    if (recorder) {
+      recorder->remove(iter->get_key());
+      recorder->insert(key, val);
+    }
+    _update(iter, key, val);
+    return;
+  }
+
+  void journal_remove(
+    const_iterator _iter,
+    delta_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    if (recorder) {
+      recorder->remove(iter->get_key());
+      return;
+    }
+    _remove(iter);
+  }
+
+  void _insert(
+    iterator iter,
+    const std::string &key,
+    const bufferlist &val) {
+    LOG_PREFIX(LogKVNodeLayout::_insert);
+    SUBDEBUG(seastore_t, "index {}", iter->get_index());
+    if (iter != iter_begin()) {
+      assert((iter - 1)->get_key() < key);
+    }
+    if (iter != iter_end()) {
+      assert(iter->get_key() > key);
+    }
+    assert(!is_overflow(key.size(), val.length()));
+    log_key_t node_key;
+    if (iter == iter_begin()) {
+      node_key.key_off = key.size() + val.length();
+      node_key.key_len = key.size();
+      node_key.val_len = val.length();
+    } else {
+      node_key.key_off = (iter - 1)->get_node_key().key_off +
+	(key.size() + val.length());
+      node_key.key_len = key.size();
+      node_key.val_len = val.length();
+    }
+    if (get_size() != 0 && iter != iter_end())
+      copy_from_local(node_key.key_len + node_key.val_len, iter + 1, iter, iter_end());
+
+    iter->set_node_key(node_key);
+    set_size(get_size() + 1);
+    iter->set_node_val(key, val);
+  }
+
+  void _update(
+    iterator iter,
+    const std::string &key,
+    const ceph::bufferlist &val) {
+    assert(iter != iter_end());
+    _remove(iter);
+    assert(!is_overflow(key.size(), val.length()));
+    _insert(iter, key, val);
+  }
+
+  void _remove(iterator iter) {
+    assert(iter != iter_end());
+    if ((iter + 1) != iter_end()) {
+      log_key_t key = iter->get_node_key();
+      copy_from_local(key.key_len + key.val_len, iter, iter + 1, iter_end());
+    }
+    set_size(get_size() - 1);
+  }
+
+  uint16_t get_entry_size(size_t ksize, size_t vsize) const {
+    return (sizeof(log_key_le_t) + ksize + vsize);
+  }
+
+  bool expect_overflow(size_t ksize, size_t vsize) const {
+    return free_space() < get_entry_size(ksize, vsize) + reserved_len;
+  }
+
+  void set_reserved_len(const uint32_t len) {
+    reserved_len = len;
+  }
+
+  uint32_t get_reserved_len() const {
+    return reserved_len;
+  }
+};
+
+
+class LogStoreManager;
+struct LogNode : LogicalChildNode, LogKVNodeLayout {
+  static constexpr extent_types_t TYPE = extent_types_t::LOG_NODE;
+  explicit LogNode(ceph::bufferptr &&ptr) : LogicalChildNode(std::move(ptr)) {
+    set_layout_buf(this->get_bptr().c_str(), this->get_bptr().length());
+    // TODO: recheck 
+    set_head_leaf(L_ADDR_NULL);
+  }
+  explicit LogNode(extent_len_t length) : LogicalChildNode(length) {}
+
+  ~LogNode() {}
+  
+  laddr_t tail_laddr = L_ADDR_NULL;
+  paddr_t tail_paddr = P_ADDR_NULL;
```

Comment:
It shouldn't need to record paddr_t's at all.  What would happen if the extent were relocated?

I'm also not sure why we need both a LogNode and a LogLeaf.  Presumably the root would simply point at the tail?